### PR TITLE
feat(component-annotate): Introduce new plugin to annotate frontend components at build-time

### PR DESCRIPTION
**Pull request by 0Calories**
Originally from this PR, splitting it into incremental smaller PRs and also making some adjustments **RE**: Abhi's suggestions:
https://github.com/getsentry/sentry-javascript-bundler-plugins/pull/464

This PR in particular adds a new package to the bundler plugins; a Babel plugin that will annotate frontend components at build-time with additional data. A lot of the source code is vendored from FullStory's React annotate plugin, but has been modified and converted to **TypeScript**. For now, the plugin only works on React, but we can extend it to other frontend frameworks / libraries in the future.

Example of an automatically annotated component in Sentry's frontend:

image

This will allow the browser SDK to pick off component names and send them to Sentry, allowing for some very useful features:

- React component names in Replays UI
- React component name search / indexing in Replays
- React component names in Breadcrumbs
- React component names in interaction transactions, grouping spans by component name

image

### CODE REVIEW

File: packages/component-annotate-plugin/README_TEMPLATE.md
Comment (AbhiPrasad) [inline packages/component-annotate-plugin/README_TEMPLATE.md:1]: Let's make this a README with some explanations about how it works?

File: packages/component-annotate-plugin/test/test-plugin.test.ts [hunk 1]
```diff
@@ -0,0 +1,2238 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+import { transform } from "@babel/core";
+import { componentNameAnnotatePlugin as plugin } from "../src/index";
+
+const BananasPizzaAppStandardInput = `import React, { Component } from 'react';
```
Comment (AbhiPrasad) [inline packages/component-annotate-plugin/test/test-plugin.test.ts:29]: can we make these separate JS files we import and assert on?

You should be able to use `fs.readFileSync` or similar.

File: packages/component-annotate-plugin/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,510 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+/**
+ * The following code is based on the FullStory Babel plugin, but has been modified to work
+ * with Sentry products
```
Comment (AbhiPrasad) [inline packages/component-annotate-plugin/src/index.ts:28]: let's make sure we detail what specifically changed between vendored code.

File: packages/component-annotate-plugin/src/constants.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+  // This module might be causing an issue preventing clicks. For safety, we won't run on this module.
+  "react-native-testfairy",
+  // This module checks for unexpected property keys and throws an exception.
+  "@react-navigation",
+];
+
+export const DEFAULT_IGNORED_ELEMENTS = [
+  "a",
+  "abbr",
+  "address",
```
Comment (AbhiPrasad) [inline packages/component-annotate-plugin/src/constants.ts:11]: Are these vendored as well? We need to add license if so.

File: packages/component-annotate-plugin/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+const webSourceFileName = "data-sentry-source-file";
+
+const nativeComponentName = "dataSentryComponent";
+const nativeElementName = "dataSentryElement";
+const nativeSourceFileName = "dataSentrySourceFile";
+
+interface AnnotationOpts {
+  native?: boolean;
+  "annotate-fragments"?: boolean;
+  ignoreComponents?: IgnoredComponent[];
```
Comment (AbhiPrasad) [inline packages/component-annotate-plugin/src/index.ts:47]: Will we expose any of these to users?

If so we should add doc strings and rename `annotate-fragments` to `annotateFragments`

File: packages/component-annotate-plugin/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,510 @@
...
+          t,
+          path,
+          path.node.id.name,
+          sourceFileNameFromState(state),
+          attributeNamesFromState(state),
+          state.opts.ignoreComponents ?? []
+        );
+      },
+      ArrowFunctionExpression(path, state) {
+        const parent = path.parent; // We're expecting a `VariableDeclarator` like `const MyComponent =`
```
Comment (AbhiPrasad) [inline packages/component-annotate-plugin/src/index.ts:80]: let's move comment above statement here.

File: packages/component-annotate-plugin/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,510 @@
...
+          !parent ||
+          !("id" in parent) ||
+          !parent.id ||
+          !("name" in parent.id) ||
+          !parent.id.name
+        ) {
+          return;
+        }
+
+        if (isKnownIncompatiblePluginFromState(state)) return;
```
Comment (AbhiPrasad) [inline packages/component-annotate-plugin/src/index.ts:91]:
```suggestion
        if (isKnownIncompatiblePluginFromState(state)) {
          return;
        }
```

ditto with other `return` cases.

**Comment by 0Calories** [comment]
Still some tweaks I need to do on this PR, but I've opened it for review for the time being

**Review by AbhiPrasad — COMMENTED** [review]
Looks good! Thanks for splitting it up :)

File: packages/component-annotate-plugin/README_TEMPLATE.md
Comment (0Calories) [reply packages/component-annotate-plugin/README_TEMPLATE.md:1]: Yep, writing that right now!

File: packages/component-annotate-plugin/src/index.ts (hunk 4 above)
Comment (0Calories) [reply packages/component-annotate-plugin/src/index.ts:47]: Not exposing this for the time being

File: packages/component-annotate-plugin/test/test-plugin.test.ts (hunk 1 above)
Comment (0Calories) [reply packages/component-annotate-plugin/test/test-plugin.test.ts:29]: I'm going to put this in the backlog for now, eventually when we extend this plugin with more functionality, I want to overhaul the test suite. I'm not entirely satisfied with the vendored tests (coverage is good but tests could be cleaner), but I want to focus on shipping this first

File: packages/component-annotate-plugin/README_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+</p>
+
+# Sentry Component Annotate Plugin
+
+[![npm version](https://img.shields.io/npm/v/@sentry/component-annotate-plugin.svg)](https://www.npmjs.com/package/@sentry/component-annotate-plugin)
+[![npm dm](https://img.shields.io/npm/dm/@sentry/component-annotate-plugin.svg)](https://www.npmjs.com/package/@sentry/component-annotate-plugin)
+[![npm dt](https://img.shields.io/npm/dt/@sentry/component-annotate-plugin.svg)](https://www.npmjs.com/package/@component-annotate-plugin)
+
+A Babel plugin that automatically annotates your output DOM with their respective frontend component names.
+This will unlock the capability to search for Replays in Sentry by component name, as well as see component names in breadcrumbs and performance monitoring.
```
Comment (AbhiPrasad) [inline packages/component-annotate-plugin/README_TEMPLATE.md:14]: Let's add minimum SDK version needed, and rename this file to be `README.md`

Also can we add a note that this is in a beta state? See https://github.com/getsentry/sentry-javascript/blob/develop/packages/bun/README.md for an example of a quick beta note.

File: packages/component-annotate-plugin/package.json [hunk 8]
```diff
@@ -0,0 +1,81 @@
+{
+  "name": "@sentry/component-annotate-plugin",
+  "version": "1.0.0",
```
Comment (AbhiPrasad) [inline packages/component-annotate-plugin/package.json:3]: This needs to align with the other version of the other packages in this repo.

**Comment by 0Calories** [comment]
Addressed your comments @AbhiPrasad

**Review by AbhiPrasad — APPROVED** [review]

**Pull request merged**